File: 1256-rank-transform-of-an-array/rank-transform-of-an-array.java
class Solution {
    public int[] arrayRankTransform(int[] arr) {
        TreeMap<Integer, Integer> rank = new TreeMap<>();
        int[] ans = new int[arr.length];

        for (int num : arr) {
            rank.put(num, 0);
        }

        int currRank = 1;
        Integer prev = null;
        for (Map.Entry<Integer, Integer> entry : rank.entrySet()) {
            if (prev != null && entry.getKey() > prev) {
                currRank++;
            }
            rank.put(entry.getKey(), currRank);
            prev = entry.getKey();
        }

        for (int i = 0; i < arr.length; i++) {
            ans[i] = rank.get(arr[i]);
        }
        return ans;
    }
}

File: 1256-rank-transform-of-an-array/rank-transform-of-an-array.py
class Solution:
    def binary_search(self, arr, target):
        left = 0
        right = len(arr) - 1

        while left <= right:
            mid = (left + right) // 2
            if target > arr[mid]:
                left = mid + 1
            elif target < arr[mid]:
                right = mid - 1
            else:
                return mid + 1

        return -1

    def arrayRankTransform(self, arr: List[int]) -> List[int]:
        arr_sorted = sorted(list(set(arr)))
        ans = []
        
        for i in range(len(arr)):
            rank = self.binary_search(arr_sorted, arr[i])
            ans.append(rank)

        return ans